jep/0000/README.adoc

```diff
@@ -0,0 +1,283 @@
+    "privateKey": "<redacted>",
+    "passphrase": "<redacted>"
+}
+```
+Note the `type` attribute, which is defined in the resource class as:
+```
+@Symbol("sshPrivateKey")
+public class SSHPrivateKeyResource extends APIResource
+```
+The `type` parameter is used for deserialization too.
+
+Next image shows the components involved in a common use of the Data API.
+`APIResource` represents the Resouces layer which isolates Model from data exposure APIs (HTTP, CLI, Configuration as Code, ...).
+`Model` boxes represent the existing model which is using the traditional `@DataBoundConstructor` (and `@DataBoundSetters`).
+`data-api-plugin` is the plugin providing the interoperability layer for consumers (it contains all needed APIs and
+tooling for both sides, consumer and model).
+
+image::components.png[Components Diagram]
+
+== Motivation
+
+[TIP]
+====
+Explain why the existing code base or process is inadequate to address the problem that the JEP solves.
+====
+
+There are currently two ways to expose data outside Jenkins: CLI and HTTP API.
+Both APIs do not guide developers on what to return and the tooling to implement them is obsolete and limited.
+Traditionally Jenkins developers have returned a serialized form of raw Jenkins model objects. The practice has serious drawbacks:
+
+* **Lack of flexibility in the API**: response messages are tied to the model (which is usually not suitable for external consumption).
+Blue Ocean is an example, a whole new HTTP API was written because the built-in Jenkins one was not enough and there was no way to adapt it
+without adapting the model too.
+* **Heterogeneous data format**: there is no Java API to guide developers on what to return, so the current public CLI and HTTP API is a
+mix of human readable format and XML of all colors.
+* **Outdated serialization tooling**: model is serialized by Stapler and based on old and limited Stapler annotations.
+* **Outdated REST API tooling**: The HTTP API is based on `hudson.model.Api` class which makes nearly impossible to write a proper REST API.
+
+== Reasoning
+
+[TIP]
+====
+Explain why particular design decisions were made.
+Describe alternate designs that were considered and related work, e.g. how the feature is supported in other systems.
+Provide evidence of consensus within the community and discuss important objections or concerns raised during discussion.
+====
+
+There are two main design decisions here:
+
+. **The use of a resources layer instead of relying on model objects directly**. The alternative is "status quo", keep exposing the model through Stapler
+and `doXX` methods (or `getDynamic` to be able to have path parameters).
+There could be an objection on the proposed resources layer: there is more code to write to expose data and sometimes resource classes will
+just mirror the model.
+It is true, but in exchange there is a huge gain on flexibility and maintainability (the model can be modified freely whilst keeping the external data
```

Comment:
+100 to this comment. Keeping the separation of your application code and your external interface ([DTO](https://en.wikipedia.org/wiki/Data_transfer_object)) cannot be understated. I have on _many many_ occasions heard the argument of "I just want to return this object [I have in my java code] to the end user, why do I need to redefine it and copy over the fields?" only to have a slight refactor to the internal class  3-months break the HTTP APIs as the response object subtly changed. 

Maintaining this separation at the class-level inside your application is the only way to safely maintain the inherent separation between internal implemtations and external interfaces that _are_ different areas of concern.